packages/cli/src/ui/hooks/useIdeTrustListener.ts

```diff
@@ -0,0 +1,53 @@
+/**
+ * @license
+ * Copyright 2025 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { useEffect, useRef, useState } from 'react';
+import type { Config } from '@google/gemini-cli-core';
+import { ideContext } from '@google/gemini-cli-core';
+
+/**
+ * This hook listens for trust status updates from the IDE companion extension.
+ * It provides the current trust status from the IDE and a flag indicating
+ * if a restart is needed because the trust state has changed.
+ */
+export function useIdeTrustListener(config: Config) {
+  const [isIdeTrusted, setIsIdeTrusted] = useState<boolean | undefined>(
+    undefined,
+  );
+  const [needsRestart, setNeedsRestart] = useState(false);
+  const initialCheckPerformed = useRef(false);
+
+  useEffect(() => {
```

Comment:
this can be made more idiomatic by using 
https://react.dev/reference/react/useSyncExternalStore

  The current implementation uses useEffect for both subscribing to a listener and managing an initial state check with a useRef flag. This can be
  simplified and made more robust by using the useSyncExternalStore hook, which is designed for subscribing to external data sources. This change
  separates the concern of syncing with the external state from the logic of determining if a restart is needed.